src/xbot/common/subsystems/ConnectionMonitorSubsystem.java

```diff
@@ -0,0 +1,43 @@
+package xbot.common.subsystems;
+
+import com.google.inject.Singleton;
+import xbot.common.command.BaseSubsystem;
+import xbot.common.logic.Latch;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.XPropertyManager;
+
+import java.util.Observable;
+import java.util.Observer;
+
+@Singleton
+public class ConnectionMonitorSubsystem extends BaseSubsystem implements Observer {
+
+    protected final DoubleProperty timeOut;
+    private final Latch connectionLatch = new Latch(true, Latch.EdgeType.FallingEdge);
+
+    private double lastPacketReceivedTimestamp = -1;
+    private double previousDisconnectionTimestamp = -1;
+
+    public ConnectionMonitorSubsystem(XPropertyManager propertyManager) {
+        log.info("Creating");
+        timeOut = propertyManager.createPersistentProperty("ConnectionMonitor TimeOut Threshold Seconds", 1.0);
+        connectionLatch.addObserver(this);
+    }
+
+    public void setLastPacketReceivedTimestamp(double currentTimestamp) {
+        if (lastPacketReceivedTimestamp != -1) {
+            connectionLatch.setValue((currentTimestamp - lastPacketReceivedTimestamp) < timeOut.get());
+        }
+        this.lastPacketReceivedTimestamp = currentTimestamp;
+    }
+
+    public double getPreviousDisconnectionTimestamp() {
```

Comment:
I don't like that this method returns `-1`... it could be confusing/unexpected to a caller. I think storing the state as `null` instead might be better. @JohnGilb Thoughts?

Reply:
I think null would be better here.

Reply:
I'm torn - I also agree that null is better in most programming, but we often program very "defensively" on the robot, trying to minimize the damage any one mistake can make, and misusing potentially-null objects is a very easy way to cause an exception.

How about sidestepping the problem entirely? Have your public method return "the number of seconds we have been disconnected." This would be 0 most of the time, and a positive double for disconnection scenarios.

In order to preserve the useful timestamp information, you could have this class be responsible for logging the exact time of disconnection.

Reply:
Additionally, although you only use this method for testing so far, I can imagine how other parts of the robot might eventually have interesting behavior in disconnection scenarios.

That being said, we still have the fundamental problem of "commands/subsystems don't run if the robot is disconnected" so this is mostly just aggressive future-proofing, and using null is probably fine for now.

Reply:
Yeah... What is the expected behavior of this code? Is it just for logging? I'm not sure any of this code will run anyway, unless we out it on its own timer on a different thread... So even logging won't ever be seen/used.

Reply:
Maybe we can add a notify() method in the base subsystem and the subsystem can override the method to do what they desire with the information. Subsystems maybe can do ConnectionMonitorSubsystem.register(Subsystem);

Reply:
We can implement more logic to detect trends and use all the other systems to see why we're disconnecting and try to fix itself.

Reply:
What I'd like to know is: what is the action that is taken in response to the disconnection? And this code requires a Boolean input indicating whether the robot is connected; where does that Boolean come from?

Separately, we must consider whether this code will ever actually run. The robot loop runs only while it is connected to the driver station; so as far as I can tell, it's just dead code.

Reply:
Lets leave this PR open and see what we can use this information for.

Reply:
But for now I see what you're saying and the code will never get executed. Only in a simulated environment like the jUnit test it looks viable.

Reply:
In theory this could be used by the robot on a reconnection - it'll notice that it didn't get any messages for a long time and just got reconnected, and at the very least we'll have a robot-side log of that particular event.

Perhaps other subsystems might want to take special action if they knew they had been suspended for a long time?

Reply:
I think both of those are interesting scenarios; the logic here would need to change though. It would be tracking time since last check rather than latching on an `isConnected` flag, and we'd probably want a notification system line @imathno mentioned where it alerted registered subsystems when operation resumed.

Alternately, we could add a logging latch to the loop frequency counter in [BaseRobot.java](https://github.com/Team488/SeriouslyCommonLib/blob/master/src/xbot/common/command/BaseRobot.java) so it logs whenever loop frequency drops below a threshold. I'm liking this latter idea the more I think about it.